### PR TITLE
Add Unicode Stripping for SMS

### PR DESCRIPTION
**Pull request by Rawi666**
Adds a functionality for stripping unicode characters for unencrypted sms messages only. Signal messages are always intact.
I grabbed UnicodeFilter class from CyanogenMod and modified it a bit and therefore I left their licence as it was.

### CODE REVIEW

**Comment by moxie0** [comment]
why?

**Comment by Rawi666** [comment]
Why I left their licence? Because I'm not sure which one I should put. I can change it if you want

**Comment by moxie0** [comment]
why do you want this functionality?

**Comment by Rawi666** [comment]
Unicode stripping is very important and popular in countries where you use non-ascii characters. If you put at least 1 non-ascii character then sms will be converted to utf16 and you can put max 70 characters in single message. This results that the application displays false information how many messages will be really sent and you are charged for multiple messages from mobile provider. Therefore aware users use this functionality to strip non-ascii characters and pay less for 1 single message

**Comment by moxie0** [comment]
https://github.com/WhisperSystems/Signal-Android/blob/master/CONTRIBUTING.md#development-ideology

I've never heard of this, and nobody has ever asked for this before. Seems unlikely that it's worth the preference?

**Comment by 2-4601** [comment]
I recall someone asking this before... Search found this old issue #1864. Still doesn't seem to be a high demand feature.

**Comment by Rawi666** [comment]
All three issues you mentioned #1864, #3217, #3257 relate to this exactly issue ;)
Trust me, this option is **very** needed in languages with non-ascii characters, but you are the boss, it's up to you.
I suspect not everyone knows that there is the discussion on github about this.
Btw. this app was mentioned in the most popular Polish portal about security https://niebezpiecznik.pl/post/ustawa-inwigilacyjna-jak-sie-bronic-przed-podgladaniem-przez-policje-i-sluzby/
In short: our government has just implementet very controversial surveillance law so the demand for this app will raise.
So please be aware that many Polish users will start using this app and here in Poland we need unicode stripping unfortunately. All keyboard suggestions always puts word with our non-ascii chars and that sucks :(

**Comment by bungabunga** [comment]
I live in Slovenia and can confirm what @Rawi666 is saying is true although not as important as it used to be, because here every smartphone user now have a plan with near unlimited amount of SMS messages included.

**Comment by JavaJens** [comment]
@Rawi666 Especially then you should not need to worry, as you should be sending encrypted messages only :wink:

**Comment by oldium** [comment]
I was just looking for an app that could send encrypted messages as well as non-encrypted SMS texts and found out this request. I also think that this feature would be really appreciated for all non-English languages (I am from the Czech Republic) if you really want to use the application for sending also non-encrypted messages via standard SMS delivery channel. I used this feature on my good old Nokia Symbian phone, which I recently replaced with a new Android phone, which now also has Unicode stripping (I switched to Cyanogenmod because of this). The Unicode stripping in my old Symbian phone was even more advanced and changed close-looking matches like á to à, ó to ò and the like. This feature would be beneficial for anybody who wants to save money (70 UTF16 characters vs 160 GSM-7bit characters in single SMS).

Typing with diacritics and removing it afterwards is needed, because all modern system keyboards offer support for writing just with diacritics during auto-completion and word correction. So having this feature is needed for proper internationalization support.

File: res/values/strings.xml [hunk 1]
```diff
@@ -999,6 +999,8 @@
     <string name="preferences__mmsc_password">MMSC Password</string>
     <string name="preferences__sms_delivery_reports">SMS delivery reports</string>
     <string name="preferences__request_a_delivery_report_for_each_sms_message_you_send">Request a delivery report for each SMS message you send</string>
+    <string name="preferences__sms_unicode_stripping">Strip unicode</string>
+    <string name="preferences__strip_all_unicode_characters">Strip all unicode characters</string>
```
Comment (moxie0) [inline res/values/strings.xml]: In Google's Messenger this is called "Simple characters only" and the summary is "Convert special characters in SMS messages you send."  Which I like better.

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 2]
```diff
@@ -126,6 +126,7 @@
 import org.thoughtcrime.securesms.util.GroupUtil;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: This change shouldn't touch anything in CoversationActivity.  Just the preferences, and SmsSendJob.

File: res/values/strings.xml (hunk 1 above)
Comment (Rawi666) [inline res/values/strings.xml]: Ok, I'll change it

File: src/org/thoughtcrime/securesms/ConversationActivity.java (hunk 2 above)
Comment (Rawi666) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: But if you move to a lower layer then with multi-character replacement like Œ->OE, Đ->Dj, etc. the app will display wrong number of characters in conversation activity and potentialy wrong number of messages to be sent. Do you really want me to move to SmsSendJob anyway?

Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: Yes, the transformation should happen in SmsSendJob.  If you want to preserve the character count, then create a new subclass of CharacterCalculator and change the base method signature to accept a charsequence rather than an int.

Comment (Rawi666) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: Ok... and wouldn't this cause storing sms message in the database in its original form (with special characters) and only send without special characters? This way the user will see his message in its original form even though the message is sent without special characters. May be a bit confusing as the sender will not see the actual replacement. Or am I wrong?

Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: What do other messengers do? Should it translate them as they're typed?

Comment (Rawi666) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: In cyanogen and in go sms pro the translation is done in the view activity so that the translated version is stored, displayed and sent consistently

Comment (Rawi666) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: I removed stripping when saving a message as a draft as this was not intended and could affect Signal messages. The rest is fine as it is as otherwise it will not work as it should

Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: Would it be better if the translation happened as you type? That way it's clear what you're sending?

Comment (Rawi666) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: I have never seen such a solution in other sms apps. This may sound like a better solution but I'm not sure if it will work for everyone for the following reasons:
- Different keyboards with different auto-completion algorithms can behave differently if the text is automatically changed as the keyboard changes the text - some keyboards still have the autocompleted word selected, some don't. This would require extensive testing.
- The translation will have to be performed each time a character is entered and that means that phones with slower CPU (or laggy roms like Samsung's...) may introduce some lag, especially when entering long messages (hard to test for me... I have Nexus 5 that is fast enough anyway I suppose). Unicode filtering uses RegExp that I think is CPU consuming in some way
- I'm not sure where would different keyboards place cursor if a user is changing (using autocompletion) text that is not at the end of the text (eg. in the middle) and the text length has changed (and it can be changed due to multi-character replacement that is in Czech language for example).
Even if it worked for me I wouldn't be sure if it worked for everyone else.
I think that all non-English users are already used to translating text just before sending messages and it's the proven solution.

The best solution in my opinion would be to implement such a feature in keyboards itself so that a user may choose in his keyboard settings that for a specific set of applications he doesn't want the keyboard to use any non-English characters, but I think that keyboards don't know what application uses it.
On the other hand that would require from the user some knowledge where to configure such a setting in his specific keyboard...

Comment (oldium) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: Usually (in SMS applications that I met) you can change the character set after you actually write the SMS (or during writing) - but before sending the message. The stored message is the one that was really sent and if you do not send it, the full character set is kept (i.e. for draft messages). The shown character count during typing respects the currently selected character set (the selected conversion).

Having this in the keyboard itself looks flexible (the application would need to tell to the keyboard about the wanted conversions), but this would also remove the option to change the character set during the typing or when you finished the typing.

**Comment by moxie0** [comment]
doesn't feel right to me in the activity, i think it should either go in the edittext or the job

**Comment by Rawi666** [comment]
OK, moved to the job

File: src/org/thoughtcrime/securesms/util/UnicodeFilter.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ * encoding (160 char limit) to 16-bit Unicode encoding (70 char limit).
+ */
+public class UnicodeFilter {
+    private static final Pattern DIACRITICS_PATTERN =
+            Pattern.compile("\\p{InCombiningDiacriticalMarks}");
+
+    public UnicodeFilter() {
+    }
+
+    public String filter(String source) {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/util/UnicodeFilter.java]: is there a reason not to make this a static method?

File: src/org/thoughtcrime/securesms/jobs/SmsSendJob.java [hunk 4]
```diff
@@ -75,6 +76,14 @@ public void onCanceled() {
     MessageNotifier.notifyMessageDeliveryFailed(context, recipients, threadId);
   }
 
+  private String getMessageBody(String body) {
+    if (TextSecurePreferences.isUnicodeStrippingEnabled(context)) {
+      return new UnicodeFilter().filter(body);
+    } else {
+      return body;
+    }
+  }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/jobs/SmsSendJob.java]: sadly this class doesn't have a serialization version id, and i'm not willing to break pending jobs for this change, so please make the change such that it doesn't break the current serialization signature

Comment (Rawi666) [inline src/org/thoughtcrime/securesms/jobs/SmsSendJob.java]: I made UnicodeFilter.filter a static method so it should be ok now. In fact there are no reasons for this method to be instance as there are no instance fields.

File: src/org/thoughtcrime/securesms/jobs/SmsSendJob.java [hunk 5]
```diff
@@ -75,6 +76,14 @@ public void onCanceled() {
     MessageNotifier.notifyMessageDeliveryFailed(context, recipients, threadId);
   }
 
+  private String getMessageBody(String body) {
+    if (TextSecurePreferences.isUnicodeStrippingEnabled(context)) {
+      return UnicodeFilter.filter(body);
+    } else {
+      return body;
+    }
+  }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/jobs/SmsSendJob.java]: still need to preserve the serialization signature

Comment (Rawi666) [inline src/org/thoughtcrime/securesms/jobs/SmsSendJob.java]: I'm not sure if we are on the same page.
Do you refer to "private static final long serialVersionUID = 0L;" inside UnicodeFilter ?
It's not needed, UnicodeFilter is not being serializer or deserialized and it's working perfectly fine as it is now (already checked).
Or do you refer to something else?

**Comment by milankni** [comment]
Thanks guys for implementing unicode stripping.

Apart from issues mentioned earlier: even that I could type w/o diacritics, such words destroy the keyboard dictionary (not only for the particular user, but also for others because some big-brother companies use crowd sourced dictionaries).

**Comment by eerozeteen** [comment]
What's about this function? Is it working? I didn't see any new options in Signal about "Use 7-bit GSM Alphabet". What should I do to make it possible?

**Comment by pmav99** [comment]
Any updates on this one? Being able to use Signal for both encrypted/unencrypted messages is a really important point and double SMS charges are surely a break dealer for many people

**Comment by Kamul-PL** [comment]
Hello!

I only wanted to admit, that I'm also looking forward for "unicode stripping" option. Are there any news in this topic? I look at all options in preferences on every Signal beta release, but unfortunately can't find this one, really important for me...

**Comment by milankni** [comment]
Hello guys,
is there any progress on this topic (i.e. the possibility to write accented chars but strip (transliterate) those to plain 7-bit ASCII before sending the unencrypted SMS)?
Paying up to three times more for monthly SMS bill does not promote the use of Signal as messaging app replacement...

**Comment by Rawi666** [comment]
I think this feature is not welcomed in Signal and therefore it stuck on pull request

**Comment by milankni** [comment]
I see, it is a pity... While I can live with a separate app for SMS (supporting this feature), the rest of my family does not seem to bother. I'll check other options available if they fit better our use case.

**Comment by Kamul-PL** [comment]
That's a pity that feature that is already done is not available to the users. Even if it is useful for many people and for sure it wouldn't bloat Signal... :/

**Comment by josepedrocorreia** [comment]
Any updates? I just started using Signal and would like to use it as my SMS app too. Being Portuguese, there are a lot of non-ASCII characters in my messages, which limits their length substantially.

**Comment by iriseecz** [comment]
Any updates? Will be implemented in the near future?

**Comment by aeharake** [comment]
> 
> 
> Unicode stripping is very important and popular in countries where you use non-ascii characters. If you put at least 1 non-ascii character then sms will be converted to utf16 and you can put max 70 characters in single message. This results that the application displays false information how many messages will be really sent and you are charged for multiple messages from mobile provider. Therefore aware users use this functionality to strip non-ascii characters and pay less for 1 single message

Could you explain what does it really strip? 
I mean if I type my name in Arabic and send it through signal to my friend, and my friend opens the normal sms app, can they view my name? or will it be only visible through the signal app?
